src/actions/receive-profile.js

```diff
@@ -38,13 +44,116 @@ export function waitingForProfileFromAddon(): Action {
   };
 }
 
-export function viewProfile(profile: Profile): Action {
-  return {
-    type: 'VIEW_PROFILE',
-    profile,
+export function viewProfile(
+  profile: Profile,
+  pathInZipFile: ?string
+): ThunkAction<void> {
+  return (dispatch, getState) => {
+    const threadIndexes = profile.threads.map((_, threadIndex) => threadIndex);
+    let hiddenThreadIndexes;
+    let selectedThreadIndex = getSelectedThreadIndexOrNull(getState());
+
+    if (selectedThreadIndex === null) {
+      // This is a new profile, it has no selected thread index.
+      hiddenThreadIndexes = _hideIdleThreads(profile);
+    } else {
+      // Pull the existing data that was URL encoded.
+      hiddenThreadIndexes = getHiddenThreads(getState())
+        // Try to use the hidden threads specified in the URL, but ensure that the indexes
+        // are correct by filtering out invalid ones.
+        .filter(index => threadIndexes.includes(index));
+    }
+
+    if (
+      // Make sure the thread index actually exists.
+      !threadIndexes.includes(selectedThreadIndex) ||
+      // And that it's not hidden.
+      hiddenThreadIndexes.includes(selectedThreadIndex)
+    ) {
+      const visibleThreads = threadIndexes
+        .filter(threadIndex => !hiddenThreadIndexes.includes(threadIndex))
+        .map(threadIndex => profile.threads[threadIndex]);
+
+      // Select either the GeckoMain [tab] thread, or the first thread in the thread
+      // order.
+      selectedThreadIndex = profile.threads.indexOf(
+        _findDefaultThread(visibleThreads)
+      );
```

Comment:
so if `_findDefaultThread` returns null, we have `selectedThreadIndex === -1` . Should we have `null` instead ?